ballet/eng/base.py

```diff
@@ -61,11 +71,137 @@ class GroupedFunctionTransformer(SimpleFunctionTransformer):
+        values will be imputed according to the mean of its group on the
+        training set, avoiding any data leakage.
+
+        .. code-block:: python
+
+           >>> from sklearn.impute import SimpleImputer
+           >>> transformer = GroupwiseTransformer(
+           ...     SimpleImputer(strategy='mean'),
+           ...     groupby_kwargs = {'by': 'name'}
+           ... )
+
+    Raises:
+        ballet.exc.Error: if handle_unknown=='error' and an unknown group is
+            encountered at transform-time.
+    """
+
+    def __init__(self,
+                 transformer,
+                 groupby_kwargs=None,
+                 handle_unknown='error'):
+        self.transformer = transformer
+        self.groupby_kwargs = groupby_kwargs
+        self.handle_unknown = handle_unknown
+
+    def _make_transformer(self):
+        if type(self.transformer) is type or callable(self.transformer):
+            return self.transformer()
+        else:
+            return sklearn.base.clone(self.transformer)
+
+    def fit(self, X, y=None, **fit_kwargs):
+        # validation on inputs
+        self.groupby_kwargs_ = self.groupby_kwargs or {}
+        if self.handle_unknown not in ['error', 'ignore']:
+            raise ValueError(
+                'Invalid value for handle_unknown: {}'
+                .format(self.handle_unknown))
+
+        # Get the groups
+        grouper = X.groupby(**self.groupby_kwargs_)
+        self.groups_ = set(grouper.groups.keys())
+
+        # Create and fit a transformer for each group
+        self.transformers_ = {}
+        for group_name, x_group in grouper:
+            transformer = self._make_transformer()
+
+            if y is not None:
+                # Extract y by integer indexing
+                inds = grouper.indices[group_name]
+                y_group = y[inds]
+                transformer.fit(x_group, y_group)
+            else:
+                transformer.fit(x_group)
+
+            self.transformers_[group_name] = transformer
+
+        return self
+
+    def transform(self, X, **transform_kwargs):
+        check_is_fitted(self, ['groups_', 'transformers_'])
+
+        def _transform(x_group, *args, **kwargs):
+            if x_group.name in self.transformers_:
+                transformer = self.transformers_[x_group.name]
+                data = transformer.transform(x_group, *args, **kwargs)
+
+                # this post-processing step is required because sklearn
+                # transform converts a DataFrame to an array. This is my
+                # best attempt so far to approximate the following:
+                # >>> result = x_group.copy()
+                # >>> result.values = data
+                # which is an error as `values` cannot be set.
+                # Unfortunately, this approach is not robust to different
+                # data types.
+                if not isinstance(x_group, pd.DataFrame):
```

Comment:
should this check be pushed out to line 176 - check that X is a pd.Df before the groupby calls?

Reply:
I will move this to 178.